runtime/registry.go

```diff
@@ -187,35 +195,88 @@ func (r *Runtime) EditInstance(ctx context.Context, inst *drivers.Instance) erro
-	r.connCache.evict(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
-	r.connCache.evict(ctx, inst.ID, inst.RepoDriver, inst.RepoDSN)
+	c, _, _ := r.OLAPDef(inst)
+	r.connCache.evict(ctx, inst.ID, c.Type, r.variables(inst.OLAPDriver, c.Configs, inst.ResolveVariables()))
+	c, _, _ = r.RepoDef(inst)
+	r.connCache.evict(ctx, inst.ID, c.Type, r.variables(inst.RepoDriver, c.Configs, inst.ResolveVariables()))
 
 	// evict catalog cache
 	r.migrationMetaCache.evict(ctx, inst.ID)
 	// query cache can't be evicted since key is a combination of instance ID and other parameters
 }
 
-func (r *Runtime) checkRepoConnection(inst *drivers.Instance) (drivers.Connection, drivers.RepoStore, error) {
-	repo, err := drivers.Open(inst.RepoDriver, map[string]any{"dsn": inst.RepoDSN}, r.logger)
+func (r *Runtime) checkRepoConnection(inst *drivers.Instance) (drivers.Handle, drivers.RepoStore, error) {
+	c, shared, err := r.RepoDef(inst)
 	if err != nil {
 		return nil, nil, err
 	}
-	repoStore, ok := repo.AsRepoStore()
+	repo, err := drivers.Open(c.Type, r.variables(c.Name, c.Configs, inst.ResolveVariables()), shared, r.logger)
+	if err != nil {
+		return nil, nil, err
+	}
+	repoStore, ok := repo.AsRepoStore(inst.ID)
 	if !ok {
 		return nil, nil, fmt.Errorf("not a valid repo driver: '%s'", inst.RepoDriver)
 	}
 
 	return repo, repoStore, nil
 }
 
-func (r *Runtime) checkOlapConnection(inst *drivers.Instance) (drivers.Connection, drivers.OLAPStore, error) {
-	olap, err := drivers.Open(inst.OLAPDriver, map[string]any{"dsn": inst.OLAPDSN}, r.logger)
+func (r *Runtime) checkOlapConnection(inst *drivers.Instance) (drivers.Handle, drivers.OLAPStore, error) {
+	c, shared, err := r.OLAPDef(inst)
+	if err != nil {
+		return nil, nil, err
+	}
+	olap, err := drivers.Open(c.Type, r.variables(c.Name, c.Configs, inst.ResolveVariables()), shared, r.logger)
 	if err != nil {
 		return nil, nil, err
 	}
-	olapStore, ok := olap.AsOLAP()
+	olapStore, ok := olap.AsOLAP(inst.ID)
 	if !ok {
 		return nil, nil, fmt.Errorf("not a valid OLAP driver: '%s'", inst.OLAPDriver)
 	}
 	return olap, olapStore, nil
 }
+
+func (r *Runtime) OLAPDef(inst *drivers.Instance) (*runtimev1.ConnectorDef, bool, error) {
+	for _, c := range inst.Connectors {
+		if c.Name == inst.OLAPDriver {
+			return c, false, nil
+		}
+	}
+	if c, err := r.ConnectorDefByName(inst.OLAPDriver); err == nil {
+		return &runtimev1.ConnectorDef{Name: c.Name, Type: c.Type, Configs: c.Configs}, true, nil
+	}
+	return nil, false, fmt.Errorf("dev error, olap connector doesn't exist")
+}
+
+func (r *Runtime) RepoDef(inst *drivers.Instance) (*runtimev1.ConnectorDef, bool, error) {
+	for _, c := range inst.Connectors {
+		if c.Name == inst.RepoDriver {
+			return c, false, nil
+		}
+	}
+	if c, err := r.ConnectorDefByName(inst.RepoDriver); err == nil {
+		return &runtimev1.ConnectorDef{Name: c.Name, Type: c.Type, Configs: c.Configs}, true, nil
+	}
+	return nil, false, fmt.Errorf("dev error, repo connector doesn't exist")
+}
```

Comment:
As hinted elsewhere – could these be replaced by something like a generic `r.Connector(instID, name)` function that generates a *resolved* connector definition (including merging relevant variables to the config payload). Which in turn could be used everywhere we connect to an instance connector

Reply:
The `registry.go` is the only place where we expect to directly construct such payloads since instance hasn't been created. Everywhere else we don't expect clients to call `r.Connector` so I didn't create such functions. There are some nuances as well like connector config is `map[string]string` and driver expects `map[string]any` so returning a connector def with all resolved variables is not straightforward as well. 
Since this is the only place we can take a call to construct such payloads(which also uses the same function that AcquireHandle internally does). Alternatively we could save instance in db first and then do all the check like valid repo, valid olap etc and delete instance if required so that we can always call `AcquireHandle`

Reply:
What if just create the instance without validating the olap and repo? It should be a rare case that they fail, and I think the error would just be returned from the first reconcile instead. Then people could update the credentials to make it work/

Reply:
Yeah that should simplify the instance API workflow as well which are slightly complicated right now.